macros/sql/deduplicate.sql

```diff
@@ -29,6 +61,44 @@
 
 {%- endmacro -%}
 
+{#
+-- For data platforms that don't support QUALIFY or an equivalent, the
+-- best we can do to ensure null safety is to use a window function +
+-- filter (which returns an extra column):
+-- https://modern-sql.com/caniuse/qualify
+#}
+{%- macro _safe_deduplicate(relation, partition_by, order_by, row_alias="rn") -%}
+
+    with row_numbered as (
+        select
+            _inner.*,
+            row_number() over (
+                partition by {{ partition_by }}
+                order by {{ order_by }}
+            ) as {{ row_alias }}
+        from {{ relation }} as _inner
+    )
+
+    select *
```

Comment:
What databases allow for `minus` or `except` syntax? I know [snowflake](https://docs.snowflake.com/en/sql-reference/operators-query#minus-except) does - that could be an option for removing the extra column. Though maybe in that case you'd just use `qualify`

Reply:
How would [minus or except](https://docs.snowflake.com/en/sql-reference/operators-query#minus-except) work to remove extra column(s)? Do you mean `select * exclude ( <col_name>, <col_name>, ... )`?

This would be the perfect solution if we could rely on it! 💡 

But it is not in the [SQL standard](https://en.wikipedia.org/wiki/ISO/IEC_9075), and the databases that don't have `qualify` are probably missing `select * exclude (...)` as well. So I don't think we'll be able to reliably use it as part of the default implementation 😢.

---

#### `select * exclude (...)`

[Snowflake](https://docs.snowflake.com/en/sql-reference/sql/select#label-select-cmd-examples-select-all-in-table-but-two) has `select * exclude`:

<img width="538" alt="image" src="https://github.com/dbt-labs/dbt-utils/assets/44704949/346cd5e5-f7dd-4aa2-8cef-24fa2b067287">

And so does [DuckDB](https://duckdb.org/docs/sql/query_syntax/select.html):

<img width="609" alt="image" src="https://github.com/dbt-labs/dbt-utils/assets/44704949/bde2d8fb-8a2b-466c-b526-ce55e072c85b">

#### `select * except (...)`

And because it's not in the standard, other databases use `except` instead of `exclude`.

[BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#select_except) uses `except`:

<img width="238" alt="image" src="https://github.com/dbt-labs/dbt-utils/assets/44704949/c0b4de08-7df9-46ea-a65b-7c58aa647ac9">

As does [Databricks](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-qry-select.html#syntax):

<img width="650" alt="image" src="https://github.com/dbt-labs/dbt-utils/assets/44704949/c006744b-3bcf-453f-b9b3-c2c248397341">

Reply:
Sorry, yes I meant `exclude`. What about using the [star](https://github.com/dbt-labs/dbt-utils#star-source) macro with the `except` argument?

Reply:
The initial implementation in [#512](https://github.com/dbt-labs/dbt-utils/pull/512/files) used the `star` macro but it was removed in https://github.com/dbt-labs/dbt-utils/pull/548.

I haven't considered the details of how we might be able to bring it back or what those implications would be.

I think we'd still need to handle the case where the `relation` is a CTE name instead of a [Relation](https://docs.getdbt.com/reference/dbt-classes#relation). That's the case that this draft PR is covering with the `row_alias` parameter. An alternative way to cover it would be a `columns` parameter like suggested [here](https://github.com/dbt-labs/dbt-utils/issues/713#issuecomment-1341066314). Allowing the end user to choose between either `row_alias` or `columns` would provide the most optionality.

Reply:
@graciegoheen your idea about using the [star](https://github.com/dbt-labs/dbt-utils#star-source) macro inspired https://github.com/dbt-labs/dbt-utils/pull/815/commits/fe03f43cc5f57d86006addfa1770ab78230cd725.

It retrieves columns similarly to `dbt_utils.star` IFF:
- relation is a Relation
- relation is not an ephemeral CTE

Otherwise, a user can pass a list of `columns` manually (d46676e98bd3e0c5f3d4e5056ff83c765720a3a0). Or they can specify a `row_alias` that is acceptable to them.